File: plugins/collector/event/networks/iptables.go
package networks

import (
	"collector/eventmanager"
	"collector/utils"
	"strings"
	"time"

	"github.com/chriskaliX/SDK"
	"github.com/chriskaliX/SDK/transport/protocol"
	"github.com/coreos/go-iptables/iptables"
)

const (
	tableFilter  = "filter"
	tableNat     = "nat"
	tableRaw     = "raw"
	tableMangble = "mangle"
)

var tableList = []string{tableFilter, tableNat, tableNat, tableMangble}
var iptablesInterval = 50 * time.Millisecond

// The iptables in OSQUERY does not work well since it only reads /proc/net/ip_tables_names
// and not works if new nftables is introduced.
type Iptables struct{}

func (Iptables) DataType() int { return 3013 }

func (Iptables) Name() string { return "iptables" }

func (Iptables) Flag() eventmanager.EventMode { return eventmanager.Periodic }

func (Iptables) Immediately() bool { return false }

func (i Iptables) Run(s SDK.ISandbox, sig chan struct{}) (err error) {
	hash := utils.Hash()
	for _, name := range tableList {
		if records, err := i.listTable(name); err == nil {
			for _, record := range records {
				rec := &protocol.Record{
					DataType:  int32(i.DataType()),
					Timestamp: utils.Clock.Now().Unix(),
					Data: &protocol.Payload{
						Fields: record,
					},
				}
				rec.Data.Fields["package_seq"] = hash
				s.SendRecord(rec)
			}
		}
	}
	return
}

func (Iptables) listTable(name string) (m []map[string]string, err error) {
	tables, err := iptables.New()
	if err != nil {
		return nil, err
	}
	chains, err := tables.ListChains(name)
	if err != nil {
		return nil, err
	}

	for _, chain := range chains {
		time.Sleep(iptablesInterval)
		rule, err := tables.List(name, chain)
		if err != nil {
			continue
		}
		stats, err := tables.Stats(name, chain)
		if err != nil {
			continue
		}

		if len(stats) > 0 {
			for _, stat := range stats {
				if len(stat) < 10 {
					continue
				}
				ipt := map[string]string{
					"table":       name,
					"chain":       chain,
					"rule":        strings.Join(rule, " "),
					"pkt":         stat[0],
					"bytes":       stat[1],
					"target":      stat[2],
					"prot":        stat[3],
					"opt":         stat[4],
					"in":          stat[5],
					"out":         stat[6],
					"source":      stat[7],
					"destination": stat[8],
					"options":     stat[9],
				}
				m = append(m, ipt)
			}
		} else {
			m = append(m, map[string]string{
				"table":       name,
				"chain":       chain,
				"rule":        strings.Join(rule, " "),
				"pkt":         "",
				"bytes":       "",
				"target":      "",
				"prot":        "",
				"opt":         "",
				"in":          "",
				"out":         "",
				"source":      "",
				"destination": "",
				"options":     "",
			})
		}
	}
	return
}

func init() { addEvent(&Iptables{}, 24*time.Hour) }

File: plugins/collector/src/event/networks/iptables.rs
/// Iptables collector — data_type 3013.
/// Reads tables via `iptables -L -n -v` / `ip6tables`.
/// Mirrors Go's event/networks/iptables.go field layout.
use std::collections::HashMap;
use std::process::Command;
use std::time::Duration;

use anyhow::Result;
use async_trait::async_trait;
use sdk::Client;

use crate::manager::{EventMode, IEvent};
use crate::event::{hash, make_record};

const DATA_TYPE: i32 = 3013;

static TABLES: &[&str] = &["filter", "nat", "mangle", "raw"];

pub struct Iptables;

#[async_trait]
impl IEvent for Iptables {
    fn name(&self)        -> &'static str { "iptables" }
    fn data_type(&self)   -> i32          { DATA_TYPE }
    fn flag(&self)        -> EventMode    { EventMode::Periodic }
    fn immediately(&self) -> bool         { false }

    async fn run(&mut self, client: &mut Client) -> Result<()> {
        let records = tokio::task::spawn_blocking(collect_iptables).await?;
        let seq     = hash();
        for mut fields in records {
            fields.insert("package_seq".into(), seq.clone());
            let _ = client.send_record(&make_record(DATA_TYPE, fields));
            tokio::time::sleep(Duration::from_millis(5)).await;
        }
        Ok(())
    }
}

/// Returns one record per rule across ipv4 + ipv6.
fn collect_iptables() -> Vec<HashMap<String, String>> {
    let mut result = Vec::new();
    for (bin, _family) in [("iptables", "4"), ("ip6tables", "6")] {
        for table in TABLES {
            // -L <chain> -n -v --line-numbers for each chain in the table
            let out = match Command::new(bin)
                .args(["-t", table, "-L", "-n", "-v", "--line-numbers", "-x"])
                .output() {
                Ok(o) if o.status.success() => o,
                _ => continue,
            };
            let text = String::from_utf8_lossy(&out.stdout);
            let mut chain = String::new();
            for line in text.lines() {
                let t = line.trim();
                if t.starts_with("Chain ") {
                    // "Chain INPUT (policy ACCEPT 0 packets, 0 bytes)"
                    chain = t.split_whitespace().nth(1).unwrap_or("").to_owned();
                    continue;
                }
                // Header lines
                if t.starts_with("num") || t.starts_with("pkts") || t.is_empty() { continue; }
                // Data line: num pkts bytes target prot opt in out source destination [options]
                let cols: Vec<&str> = t.split_whitespace().collect();
                if cols.len() < 10 { continue; }
                let mut m: HashMap<String, String> = HashMap::new();
                m.insert("table".into(),       table.to_string());
                m.insert("chain".into(),       chain.clone());
                m.insert("pkt".into(),         cols[1].to_owned());
                m.insert("bytes".into(),       cols[2].to_owned());
                m.insert("target".into(),      cols[3].to_owned());
                m.insert("prot".into(),        cols[4].to_owned());
                m.insert("opt".into(),         cols[5].to_owned());
                m.insert("in".into(),          cols[6].to_owned());
                m.insert("out".into(),         cols[7].to_owned());
                m.insert("source".into(),      cols[8].to_owned());
                m.insert("destination".into(), cols[9].to_owned());
                m.insert("options".into(),     cols.get(10..).map(|s| s.join(" ")).unwrap_or_default());
                m.insert("rule".into(),        t.to_owned());
                result.push(m);
            }
        }
    }
    result
}
